databricks/sdk/_base_client.py

```diff
@@ -237,9 +259,15 @@ def _perform(self,
                                          stream=raw,
                                          timeout=self._http_timeout_seconds)
         self._record_request_log(response, raw=raw or data is not None or files is not None)
+
         error = self._error_parser.get_api_error(response)
         if error is not None:
+            # If the request body is a seekable stream, rewind it so that it is ready
+            # to be read again in case of a retry.
+            if self._is_seekable_stream(data):
+                data.seek(initial_data_position)
```

Comment:
Nit: we're doing seek during the last retry attempt (after handling failure), in which case we'll not use this stream after seeking. If seek is a costly operation, this may matter.

Reply:
Agreed that this is a weakness of this solution. The alternative is to implement the unwind logic within the retry logic. Unfortunately, the current code (with the decorator) does not make this easy. 

My recommendation would be to proceed with this PR first to fix the bug, and clean-up in a follow-up PR focused on optimization. Please feel free to disagree.

Reply:
This just feels a bit awkward to me although I don't have a strong example. Could a customer rely on the resulting stream position in their code?

For the solution, we could pass a lambda to the retry function that would call it before each retry attempt. This lambda could even return e.g. false in case stream is not seekable, so that retry logic bubbles up the error.